src/ui/RelevanceInspector/AvailableFieldsTable.ts

```diff
@@ -0,0 +1,271 @@
+
+        list.append(listItem.el);
+      });
+    }
+
+    return list;
+  }
+
+  public refresh() {
+    return true;
+  }
+}
+
+interface IFilterModel {
+  type: 'equals' | 'notEqual' | 'startsWith' | 'endsWith' | 'contains' | 'notContains';
+  filter: string;
+}
+
+interface ISortModel {
+  colId: string;
+  sort: 'asc' | 'desc';
+}
+
+export class AvailableFieldsDatasource implements agGridModule.IDatasource {
+  private rowsData: any[] = [];
+  constructor(public dataSource: Record<string, ITableDataSource>[]) {
+    this.rowsData = map(this.dataSource, source => {
+      const merged: any = {};
+
+      const extractContent = (value: ITableDataSource & agGridModule.ColGroupDef, key: string) => {
+        if (value.content != null) {
+          merged[key] = value.content;
+        } else if (value.children) {
+          each(value.children, (child: any) => {
+            each(child, extractContent);
+          });
+        }
+      };
+
+      each(source, extractContent as any);
+      return merged;
+    });
+  }
+
+  public getRows(params: agGridModule.IGetRowsParams) {
+    const filtered = this.filter(this.rowsData, params.filterModel);
+    const filteredAndSorted = this.sort(filtered, params.sortModel);
+
+    params.successCallback(filteredAndSorted.slice(params.startRow, params.endRow), filteredAndSorted.length);
+  }
+
+  private filter(rows: any[], filterModel: Record<string, IFilterModel>) {
+    each(filterModel, (value, key) => {
+      switch (value.type) {
+        case 'equals':
+          rows = where(rows, { [key]: value.filter });
+          break;
+
+        case 'notEqual':
+          rows = reject(rows, possibleReturn => {
+            return possibleReturn[key] == value.filter;
+          });
+          break;
+
+        case 'startsWith':
+          rows = filter(rows, possibleReturn => {
+            return possibleReturn[key].toLowerCase().indexOf(value.filter) == 0;
+          });
+          break;
+
+        case 'endsWith':
+          rows = filter(rows, possibleReturn => {
+            return possibleReturn[key].toLowerCase().indexOf(value.filter) == possibleReturn[key].length - value.filter.length;
+          });
+          break;
+
+        case 'contains':
+          rows = filter(rows, possibleReturn => {
+            return possibleReturn[key].toLowerCase().indexOf(value.filter) != -1;
+          });
+          break;
+
+        case 'notContains':
+          rows = filter(rows, possibleReturn => {
+            return possibleReturn[key].toLowerCase().indexOf(value.filter) == -1;
+          });
+          break;
+      }
+    });
+
+    return rows;
+  }
+
+  private sort(rows: any[], sortModels: ISortModel[]) {
+    if (sortModels && sortModels[0]) {
+      const sortModel = sortModels[0];
+      return rows.sort((first, second) => {
+        const firstValue = first[sortModel.colId];
+        const secondValue = second[sortModel.colId];
+        if (isNaN(firstValue) || isNaN(secondValue)) {
```

Comment:
They are all non empty in practice because everything that gets rendered in the table passes through a generic rendering function will will change null/emptry value to "-- NULL --" string for readability purpose.

In theory though, you are right, I'll add defensive coding here.